CommonTools/RecoAlgos/interface/FKDTree.h

```diff
@@ -0,0 +1,386 @@
+#include "FKDPoint.h"
+#include "FQueue.h"
+
+
+namespace{
+    const std::array<unsigned int,32> MultiplyDeBruijnBitPosition{{0, 9, 1, 10, 13, 21, 2, 29, 11, 14, 16, 18, 22, 25, 3, 30,
+  8, 12, 20, 28, 15, 17, 24, 7, 19, 27, 23, 6, 26, 5, 4, 31}};
+    unsigned int ilog2 (unsigned int v)
+	    {
+	    v |= v >> 1; // first round down to one less than a power of 2 
+	    v |= v >> 2;
+	    v |= v >> 4;
+	    v |= v >> 8;
+	    v |= v >> 16;
+	    return MultiplyDeBruijnBitPosition[(unsigned int)(v * 0x07C4ACDDU) >> 27];
+
+	    }
+
+
+
+}
+
+
+
+
+template<class TYPE, unsigned int numberOfDimensions>
+class FKDTree
+{
+
+public:
+
+
+FKDTree()
+	{
+		theNumberOfPoints = 0;
+		theDepth = 0;
+
+	}
+
+
+
+	void push_back(const FKDPoint<TYPE, numberOfDimensions>& point)
+	{
+
+		thePoints.push_back(point);
+
+
+	
+	}
+
+	void push_back(FKDPoint<TYPE, numberOfDimensions> && point)
+	{
+
+		thePoints.emplace_back(point);
+
+
+	}
+
+	template <typename ... Args>
+	void emplace_back(Args&&... args) {
+	   push_back(FKDPoint<TYPE, numberOfDimensions>(std::forward<Args>(args)...));
+	}
+
+	bool empty()
+	{
+		return theNumberOfPoints==0;
+	}
+
+
+	void search(
+			const FKDPoint<TYPE, numberOfDimensions>& minPoint,
+			const FKDPoint<TYPE, numberOfDimensions>& maxPoint,
+			std::vector<unsigned int>& foundPoints)
+	{
+		FQueue<unsigned int> indicesToVisit(16);
+		indicesToVisit.push_back(0);
+		unsigned int index;
+		bool intersection;
+		int dimension;
+		unsigned int numberOfindicesToVisitThisDepth;
+		int numberOfSonsToVisitNext;
+		unsigned int firstSonToVisitNext;
+
+		for (unsigned int depth = 0; depth < theDepth + 1; ++depth)
+		{
+
+			dimension = depth % numberOfDimensions;
+			numberOfindicesToVisitThisDepth = indicesToVisit.size();
+			for (unsigned int visitedindicesThisDepth = 0;
+					visitedindicesThisDepth < numberOfindicesToVisitThisDepth;
+					visitedindicesThisDepth++)
+			{
+
+				index = indicesToVisit[visitedindicesThisDepth];
+				intersection = intersects(index, minPoint, maxPoint, dimension);
+				firstSonToVisitNext = leftSonIndex(index);
+            
+
+				if (intersection)
+				{
+					if (is_in_the_box(index, minPoint, maxPoint))
+					{
+										
+						foundPoints.emplace_back(theIds[index]);
+					}
+					numberOfSonsToVisitNext = (firstSonToVisitNext < theNumberOfPoints)
+                 		 + ((firstSonToVisitNext + 1) < theNumberOfPoints);
+				}
+				else
+				{
+
+
+					firstSonToVisitNext += (theDimensions[dimension][index]
+							< minPoint[dimension]);
+
+              		numberOfSonsToVisitNext = std::min((firstSonToVisitNext< theNumberOfPoints)
+                  		+ ((firstSonToVisitNext + 1) < theNumberOfPoints) , 1);
+				}
+
+				for (int whichSon = 0; whichSon < numberOfSonsToVisitNext;
+						++whichSon){
+               
+					indicesToVisit.push_back(firstSonToVisitNext + whichSon);
+			      }
+			}
+
+			indicesToVisit.pop_front(numberOfindicesToVisitThisDepth);
+		}
+	}
+
+
+	void build()
```

Comment:
The total duplication of the `build()` function just to change the source of the points vector is unnecessary. This function only needs to exist once.